### PR TITLE
Add BSplineSE3 class.

### PR DESCRIPTION
**Pull request by myeatman-bdai**
This adds a class to represent SE3 trajectories using a b-spline (using scipy's b-spline class under the hood). A small example is included, with a visualization.

Fixes a minor bug with the animation functionality.

### CODE REVIEW

File: spatialmath/spline.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+from scipy.interpolate import BSpline
+from spatialmath import SE3
+import numpy as np
+import matplotlib.pyplot as plt
+from spatialmath.base.transforms3d import tranimate, trplot
+
+class BSplineSE3:
+    """ A class to parameterize a trajectory in SE3 with a 6-dimensional B-spline.
+
+    The SE3 control poses are converted to se3 twists (the lie algebra) and a B-ppline
```
Comment (amessing-bdai) [inline spatialmath/spline.py:20]:
```suggestion
    The SE3 control poses are converted to SE3 twists (the lie algebra) and a B-spline 
```

File: spatialmath/spline.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    The SE3 control poses are converted to se3 twists (the lie algebra) and a B-ppline 
+    is created for each dimension of the twist, using the corresponding element of the twists 
+    as the control point for the spline.  
+
+    For detailed information about B-splines, please see this wikipedia article. 
+    https://en.wikipedia.org/wiki/Non-uniform_rational_B-spline
+    """
+
+    def __init__(self, control_poses: list[SE3], degree: int = 3, knots: Optional[list[float]] = None) -> None:
```
Comment (amessing-bdai) [inline spatialmath/spline.py:28]:
```suggestion
    def __init__(self, control_poses: List[SE3], degree: int = 3, knots: Optional[List[float]] = None) -> None:
```

You need to get these from typing to support python 3.7/3.8

File: spatialmath/spline.py (hunk 1 above)
Comment (myeatman-bdai) [reply spatialmath/spline.py:20]: the "little se3" was intentional , the ppline was not

File: spatialmath/spline.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+"""
+
+from typing import Any, Optional
+from scipy.interpolate import BSpline
+from spatialmath import SE3
+import numpy as np
+import matplotlib.pyplot as plt
+from spatialmath.base.transforms3d import tranimate, trplot
+
+class BSplineSE3:
```
Comment (amessing-bdai) [inline spatialmath/spline.py:17]: You'll need to add this to the `__init__.py`

File: spatialmath/spline.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    def __call__(self, t: float) -> SE3:
+        """ Returns pose of spline at t.
+
+        t: Normalized time value [0,1] to evaluate the spline at.
+        """
+        twist = np.hstack([spline(t) for spline in self.splines])
+        return SE3.Exp(twist)
+    
+    def visualize(self, num_samples: int, repeat: bool = False) -> None:
```
Comment (amessing-bdai) [inline spatialmath/spline.py:62]: I think in the other classes this is called plot. Can you rename and add the 
```
bounds: Optional[ArrayLike] = None,
        ax: Optional[plt.Axes] = None,
        **kwargs,
```
that the others have?

File: spatialmath/spline.py (hunk 3 above)
Comment (myeatman-bdai) [reply spatialmath/spline.py:17]: :+1:

File: spatialmath/spline.py (hunk 2 above)
Comment (myeatman-bdai) [reply spatialmath/spline.py:28]: > :+1:

File: spatialmath/spline.py (hunk 1 above)
Comment (jcao-bdai) [inline spatialmath/spline.py]: typo: B-ppline -> B-spline

File: spatialmath/spline.py (hunk 4 above)
Comment (myeatman-bdai) [reply spatialmath/spline.py:62]: I can definitely rename it, but I don't think the API for things called "plot" and "animate" is super consistent throughout the code base. I'll give it shot.

Comment (amessing-bdai) [reply spatialmath/spline.py:62]: Ah, that is unfortunate. I saw a couple that were and was hopeful

File: spatialmath/spline.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        # a matrix where each row is a control pose as a twist
+        # (so each column is a vector of control points for that dim of the twist)
+        self.control_pose_matrix = np.vstack([np.array(element.twist()) for element in control_poses]) 
+        
+        self.degree = degree 
+        
+        if knots is None:
+            knots=np.linspace(0,1,len(control_poses)-2,endpoint=True)
+            knots=np.append([0,0,0],knots) # ensures the curve starts on the first control pose
+            knots=np.append(knots,[1,1,1])  # ensures the curve ends on the last control pose
```
Comment (jcao-bdai) [inline spatialmath/spline.py:49]: it seems that the length of the appended series, [0, 0, 0] and [1, 1, 1], should be variable, `[0.0] * degree` and `[1.0] * degree` (and avoid implicit int->float conversion)

File: spatialmath/spline.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+        t: Normalized time value [0,1] to evaluate the spline at.
+        """
+        twist = np.hstack([spline(t) for spline in self.splines])
+        return SE3.Exp(twist)
+    
+    def visualize(self, num_samples: int, repeat: bool = False) -> None:
+        """ Displays an animation of the trajectory with the control poses.
+        """
+        out_poses = [self(i) for i in np.linspace(0,1,num_samples)]
```
Comment (jcao-bdai) [inline spatialmath/spline.py:65]: super nit: i -> t

File: spatialmath/spline.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        *SE3.Ry(e/8 * np.pi) for e in range(0,8)
+    ]
+    
+    spline = BSplineSE3(control_poses, degree)
+    spline.visualize(num_samples=100, repeat=True)
+    
+    
+
+if __name__ == "__main__":
+    main()
```
Comment (jcao-bdai) [inline spatialmath/spline.py:92]: nit: add a line at the end of file

**Review by jcao-bdai — APPROVED** [review]
LGTM. would it help to add some tests?

File: spatialmath/spline.py (hunk 5 above)
Comment (myeatman-bdai) [reply spatialmath/spline.py:49]: That's a good catch

File: spatialmath/spline.py (hunk 6 above)
Comment (myeatman-bdai) [reply spatialmath/spline.py:74]: :+1:

**Review by jcao-bdai — APPROVED** [review]

**Review by jcao-bdai — APPROVED** [review]

**Comment by myeatman-bdai** [comment]
I'm going to merge and come back and fix these failing tests later.

**Pull request merged**